Refactor image preprocessing operations to allow new operations addition

juliabeliaeva opened the pull request:
Fixes #208.

zaleslaw commented on dataset/src/test/kotlin/org/jetbrains/kotlinx/dl/dataset/preprocessor/PreprocessingFinalShapeTest.kt:
```diff
@@ -0,0 +1,134 @@
+import org.junit.jupiter.api.Test
+import java.io.File
+
+class PreprocessingFinalShapeTest {
```
What do you think about adding tests for comparing preprocessed images with earlier preprocessed and saved to the resources? Is it too difficult to add?

zaleslaw commented on dataset/src/main/kotlin/org/jetbrains/kotlinx/dl/dataset/preprocessor/image/Resize.kt:
```diff
@@ -33,7 +33,12 @@ public class Resize(
-    override fun apply(image: BufferedImage, inputShape: ImageShape): Pair<BufferedImage, ImageShape> {
+
+    override fun getOutputShape(inputShape: ImageShape?): ImageShape {
+        return ImageShape(outputWidth.toLong(), outputHeight.toLong(), 3)
```
Could it be applied to the 1-channel images?

juliabeliaeva replied:
I think so, however the output will have 3 channels anyway. Ideally we'd want to keep the number of channels in the image. So `Resize#getOutputShape` should return `ImageShape(outputWidth.toLong(), outputHeight.toLong(), null)` (we need to make `ImageShape#channels` nullable for that).

zaleslaw commented on dataset/src/main/kotlin/org/jetbrains/kotlinx/dl/dataset/preprocessor/image/Loading.kt:
```diff
@@ -16,7 +16,7 @@ import java.nio.file.Path
 
 /**
- * This image preprocessor defines the Loading operation.
+ * This class defines the Loading operation.
```
Highly likely it should be moved to the first-level in the DSL like

   loadFile {
                    pathToData = File("test.jpg")
                    imageShape = ImageShape(20, 20, 3)
                    colorMode = ColorOrder.BGR
                }
transformImage {
                resize {
                    outputWidth = 100
                    outputHeight = 100
                    interpolation = InterpolationType.NEAREST
                }
            }

And also we could add loadFile/loadDirectories ops or 
load {
      file {
      }
}

load {
      directory {
      }
}

load {
      database {
      }
}

zaleslaw replied:
But it could be a following ticket of course, it's not a part of this refactoring

zaleslaw commented on dataset/src/main/kotlin/org/jetbrains/kotlinx/dl/dataset/preprocessor/image/ImagePreprocessor.kt:
```diff
@@ -11,13 +11,19 @@ import java.awt.image.BufferedImage
-     * @return Pair <new image; new shape>.
+     * @return output image shape
      */
-    public fun apply(image: BufferedImage, inputShape: ImageShape): Pair<BufferedImage, ImageShape>
```
So, I had yet two ideas, but didn't implement yet
1) return pair, but calculate outputShape by inputShape for each Preprocessor
2) return new object of class OurGreatImage (with fields BufferedImage and ImageShape), later it's easy to add these prepocessors like extension functions to the OurGreatImage for example

Let's discuss both approaches?

juliabeliaeva replied:
Could you explain why a separate shape is needed? `BufferedImage` has this information already, so it seems a bit redundant to me. Not to mention that because of the nullable fields using `ImageShape` might be awkward in this case.

zaleslaw commented on dataset/src/main/kotlin/org/jetbrains/kotlinx/dl/dataset/preprocessor/image/ImagePreprocessing.kt:
```diff
@@ -21,37 +20,16 @@ public class ImagePreprocessing {
+    internal val operations = mutableListOf<ImagePreprocessor>()
+
+    public fun addOperation(operation: ImagePreprocessor) {
+        operations.add(operation)
```
Yeah, thanks, I'd really like to see the removal of these functions, no needs now

zaleslaw commented on dataset/src/main/kotlin/org/jetbrains/kotlinx/dl/dataset/preprocessor/image/Cropping.kt:
```diff
@@ -29,22 +30,27 @@ public class Cropping(
-                channels = inputShape.channels
-            )
+
+    override fun getOutputShape(inputShape: ImageShape?): ImageShape? {
```
Looks like all transformation should have implementation and ability to return/control shape.

zaleslaw commented on dataset/src/main/kotlin/org/jetbrains/kotlinx/dl/dataset/preprocessor/Preprocessing.kt:
```diff
@@ -51,45 +49,25 @@ public class Preprocessing {
-        if (imagePreprocessingStage.isSaveInitialized) {
-            imagePreprocessingStage.save.imageToFile(filename, image, shape)
+        for (operation in imagePreprocessingStage.operations) {
+            if (operation is Save) {
```
Looks like we have two sort of transformation: intermediate and start/final.
What do you think about Save op? Is it an imageTransformation or we should have a separate block for it at the top level?
load {
}
transformImage {
}
save {
}

zaleslaw replied:
But on other side, what if we want to save checkpoints after each imageTransformation (or display it)

juliabeliaeva replied:
> But on other side, what if we want to save checkpoints after each imageTransformation (or display it)

Yes, that what I was thinking.

One of the possible ways it could look like:

```
resize {
    outputWidth = 400
    outputHeight = 400
    interpolation = InterpolationType.NEAREST
    save {
        dirLocation = preprocessedImagesDirectory
    }
}
```

zaleslaw reviewed (changes requested):
So, I'd really like your job in this PR; I wrote a few notes for you and suggested discussing it.

YokiToki commented on dataset/src/main/kotlin/org/jetbrains/kotlinx/dl/dataset/preprocessor/image/Save.kt:
```diff
@@ -24,7 +24,7 @@ public class Save(
-    internal fun imageToFile(filename: String, image: BufferedImage, shape: ImageShape): File {
+    internal fun imageToFile(filename: String, image: BufferedImage): File {
         val outputFile: File = if (dirLocation!!.isDirectory) {
             File("${dirLocation}\\${filename}")
```
How about fix this line `File.separator`? This is not correct for unix systems now.

zaleslaw replied:
Many thanks for the comment, do you have any suggestions for quick fix?

YokiToki replied:
I think there is a problem with declaring dirLocation as nullable. Passing null will just throw a NullPointerException. And I want to note that if the value of `dirLocation` is not a directory, then each processing result will overwrite the file.

juliabeliaeva replied:
> How about fix this line with File.separator? This is not correct for unix systems now.

I think this fix should be done as a separate pull request.

zaleslaw replied:
@YokiToki we are going to split into two parts the logic of loading the whole directory and one file, and thanks for the note about the Unix system and possible overwrinting

juliabeliaeva commented:
I pushed some fixes for this PR:
1. Defined output shapes for all operations.
2. Moved `load` out of `transformImage`. I did not split it to `loadFile/loadDirectory` as it turned into another major refactoring and I think it should be done as a separate PR.
3. Moved `save` to individual operations.
4. Added a couple of tests. I thought that perhaps adding more images to the repository may not be optimal, so I've written some simple in-memory tests.

juliabeliaeva commented:
Rebased on latest master to resolved conflicts.
